### ISSUE TITLE
payment.create was true but response was INSTRUMENT_DECLINED

### ISSUE THREAD

**Issue by bobbus** [paypal/PayPal-Ruby-SDK#234]
### General Information

- Mode (Sandbox/Live): LIVE
- PayPal-Debug-ID(s) (from any logs): 20dcd4f71107c
- Version of Ruby used: 2.1.7
- SDK Version: 1.4.6

### Issue Description

Problem was that `payment.create` was true but the transaction was refused, see log below : 

````
Response[400]: Bad Request, Duration: 2.293s
Response.body={"name":"INSTRUMENT_DECLINED","message":"The instrument presented was either declined by the processor or bank, or it can't be used for this payment.","information_link":"https://developer.paypal.com/docs/api/#INSTRUMENT_DECLINED","debug_id":"20dcd4f71107c","processor_response":{"avs_code":"Y","cvv_code":"N","response_code":"0051"}} Response.header={"date"=>["Fri, 09 Sep 2016 17:44:23 GMT"], "server"=>["Apache"], "proxy_server_info"=>["host=dcg12javapapi8768.dcg12.slc.paypalinc.com;threadId=1177"], "paypal-debug-id"=>["20dcd4f71107c", "20dcd4f71107c"], "correlation-id"=>["20dcd4f71107c"], "content-language"=>["*"], "connection"=>["close", "close"], "http_x_pp_az_locator"=>["dcg12.slc"], "set-cookie"=>["X-PP-SILOVER=name%3DLIVE3.API.1%26silo_version%3D880%26app%3Dplatformapiserv%26TIME%3D4160016983%26HTTP_X_PP_AZ_LOCATOR%3Ddcg12.slc; Expires=Fri, 09 Sep 2016 18:14:25 GMT; domain=.paypal.com; path=/; Secure; HttpOnly", "X-PP-SILOVER=; Expires=Thu, 01 Jan 1970 00:00:01 GMT"], "vary"=>["Authorization"], "content-length"=>["335"], "content-type"=>["application/json"]}
````

**Comment by lcoq** [comment]
:+1:

**Comment by davyfitz** [comment]
Hey, I've been having issues with payment.create returning true for 400 Bad Request errors and tracked it down to an issue with the http_call method, I've submitted a pull request:

https://github.com/paypal/PayPal-Ruby-SDK/pull/239

**Comment by pp-randy** [comment]
I apologize for the late response.  If this is a LIVE integration issue and happening for all your payments please contact PayPal Technical Support here:  https://www.paypal-techsupport.com/app/ask

If this is one buyer seeing an issue they should contact PayPal Customer Support for more information about the decline here: https://www.paypal.com/help

**Comment by davyfitz** [comment]
We're having the same issue sporadically, when the response is in the following format it is handled correctly:

{"error"=>{"name"=>"VALIDATION_ERROR", "details"=>[{"field"=>"payer.funding_instruments[0].credit_card.number", "issue"=>"Value is invalid"}], "message"=>"Invalid request - see details", "information_link"=>"https://developer.paypal.com/docs/api/#VALIDATION_ERROR", "debug_id"=>"877b747d3fc9c"}}

However it seems we're getting some response's similar to Bobbus where there is no "error" attribute, and in this case the payment.create method returns true. The lack of "error" attribute means that the call to the "success?" method returns true.

From data_types.rb:

        def create()
          path = "v1/payments/payment"
          response = api.post(path, self.to_hash, http_header)
          self.merge!(response)
          success?
        end

        def success?
          @error.nil?
        end

**Comment by bobbus** [comment]
@pp-randy : Hello Randy,
Thanks for your answer. The problem only happen once with a live transaction. Should we contact paypal ? I believe the cause of the problem was what @davyfitz described in the comment above.
Thanks for your help

**Comment by pp-randy** [comment]
@bobbus it happened once but not for every transaction?  The error you posted indicates the funding source was declined.  That seems different than the error from @davyfitz.

The error from @davyfitz looks like a validation_error and should have a different cause.

In this context this is my understanding of both errors.

- INSTRUMENT_DECLINED  - Valid Credit Card Number, but was Declined.

- VALIDATION_ERROR - Not a Valid Credit Card number.

Does your integration work with a known good credit card number?

**Comment by bobbus** [comment]
@pp-randy : Yes it happen once and yes our integration works with proper payment information.

The problem is not that the transaction was declined and the problem encoutered by @davyfitz is not the Validation error. 

The problem with this gem is : 

> `payment.create` was true but the transaction was refused

And @davyfitz probably detailed it better : 

> we're getting some response's similar to Bobbus where there is no "error" attribute, and in this case the payment.create method returns true. The lack of "error" attribute means that the call to the "success?" method returns true

Please see his comments above.

This should be fixed on this library or in the paypal API response to ensure we never hit the case when payment is declined but `payment.create` is `true`.

**Comment by pp-randy** [comment]
@bobbus thank you for the clarification, I will bring this to the attention of the SDK engineering team.

**Comment by sdcoffey** [comment]
Picking this up today, thanks for the info @bobbus and @davyfitz
